<html><header><title>How to override DataPortalException (or is there a better way to do what we need)</title></header><body><div style='float:right'><a href='http://cslanet.com'><img src='https://github.com/MarimerLLC/csla/raw/master/Support/Logos/csla%20win8_compact_s.png'/></a></div><p><h1>How to override DataPortalException (or is there a better way to do what we need)</h1><span style='font-size:small'>Old forum URL: forums.lhotka.net/forums/t/12806.aspx</span></p><hr><div style='padding:0 15 3 15;background-color:powderblue'><h2>SethSpearman posted on Thursday, December 11, 2014</h2><p>Hello,</p>
<p>Our shop makes extensive use of CSLA.&nbsp; We have written asp.net mvc, winforms, web services, etc all using CSLA.</p>
<p>One of the things that we are really struggling with is how to get back useful debugging information when an error occurs.&nbsp; The thing that makes it complicated is that our a single application can pass through multiple application server &quot;layers&quot; to do what it needs.</p>
<p>For example, (and this is a real example), we have a winforms application that makes a call to a remote data portal.&nbsp; Sometimes (ordering related), the data portal makes a call to a webservice on another machine.&nbsp; The webservice makes another remote data portal call to a server inside our firewall.&nbsp; </p>
<p>We have another scenario which is similar but in this case the UI is an asp.net mvc application and it hops to a diferent web service (REST based) but it calls the same final data portal as the above example.</p>
<p>My &quot;holy grail&quot; would be to get back object state and stacktrace information for ALL of the hops and log into a database somewhere.&nbsp; The user&nbsp; would just see a GUID or something that says (call tech support with this number) and we would be able to look it up and the log&nbsp; would be have all of the information across all hops. </p>
<p>Right now we have to fire up a bunch of Visual Studio environments and point the data portal to production data to replicate the issues.&nbsp; </p>
<p>I was thinking if we could override DataPortalException and add a GUID and then use implement IDataPortalExceptionInspector and change all of our configurations to call a custom exception that it would be the beginning of solving our problem.&nbsp; But I am not able to get something like that to work so far (because the client side DataPortalException does not know how to deserialize the GUID property of the new exception type).&nbsp; That was one of many errors I have encountered to try to get this to work.</p>
<p>Is there an easier way to do it?&nbsp; Or am I on the right track?&nbsp; I am not an engineer/architect so if you could give me an overview of the best solution that would be great.</p>
<p>We are testing this on the VERY NEWEST version of CSLA (4.5.601)&nbsp; and .net 4.5.</p>
<p>Seth</p>
<p>&nbsp;</p>
<p>&nbsp;</p></div><div style='padding:0 15 3 15;background-color:powderblue'><h2>skagen00 replied on Friday, December 12, 2014</h2><p>We use the enterprise library for exception logging, and here&#39;s what we end up doing.</p>
<p>On our base objects (the layer right on BusinessBase&lt;T&gt; for example), we do a try/catch within the DP_Fetch, DP_Create, etc.</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /// &lt;summary&gt;<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /// Retrieves an existing business object.<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /// &lt;/summary&gt;<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /// &lt;param name=&quot;criteria&quot;&gt;<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /// An &lt;see cref=&quot;Object&quot;/&gt; containing information necessary for retrieval.<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; /// &lt;/param&gt;<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; protected virtual void DataPortal_Fetch(object criteria) <br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; {<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; try<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; {<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; // Add the criteria to the debug objects collection.<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; OurProductLocalContext.AddDebugObject(criteria);</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; using (var dalFactory = DalFactory.GetManager())<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; {<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; var dal = dalFactory.GetProvider&lt;TData&gt;(typeof(T));<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; TData data = dal.Fetch(criteria);<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; if (data != null) FetchCore(data);<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; catch (Exception e)<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; {<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; OurProductEnterpriseLibraryUtility.HandleException(e, OurProductExceptionPolicy.ReleaseStandardPolicy);<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }</p>
<p>We also add the current object during the dataportal invocation to this &quot;context&quot; object.</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; protected override void DataPortal_OnDataPortalInvoke(DataPortalEventArgs e)<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; {<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; OurProductLocalContext.AddDebugObject(this); // Add the current object to the debug objects collection</p>
<p>&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; base.DataPortal_OnDataPortalInvoke(e);<br />&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; }</p>
<p>This context object happens to live in Csla.ApplicationContext.LocalContext.</p>
<p>We have a database trace listener attached, which ends up logging a nice XML representation of the objects to the database - because we have all the objects we care about in the LocalContext, we can iterate over them and write them out.</p>
<p>Truthfully, it haven&#39;t been used as much as I thought it would have been when I created it, but it does successfully give you the state of the objects at the time of the error. </p>
<p>It can be very verbose for certain objects, so it does require some maintenance. </p>
<p>Attached some of the key methods I use to write out a collection of objects to XML...</p></div><div style='padding:0 15 3 15;background-color:powderblue'><h2>SethSpearman replied on Friday, December 12, 2014</h2><p>skagen,</p>
<p>Thanks for your reply.&nbsp; It has given me some things to try.&nbsp; </p>
<p>One difference between your suggestion and what I was thinking how it would be done is that your&nbsp; are actually doing the logging on the server side.&nbsp; </p>
<p>One nice about doing things on the server side is that I have a reference to the BusinessObject property.&nbsp; I have noticed that on the client side the DataPortalException.BusinessObject is always null.&nbsp; The documentation seems to indicate that the BusinessObject property should contain the business object at the time of the exception.&nbsp; </p>
<p>But, again, for me it is null.&nbsp; Do you know why it might be null.&nbsp; Or is that why you are doing the work on the server side?</p>
<p>One other thing...is the reason you use the Enterprise Library ... does it automatically serialize the object into a format that can be persisted (like xml...per your post).&nbsp; I am struggling with how to persist the exception details.</p>
<p>Seth</p></div><div style='padding:0 15 3 15;background-color:powderblue'><h2>skagen00 replied on Friday, December 12, 2014</h2><p>Ah - sorry I guess the type of application matters a bit now doesn&#39;t it, along with what the desires are for logging.&nbsp; Mine is for a Silverlight application and the logging I presented is just for the DP calls that happen against the server. </p>
<p>For logging client side there is definitely less contextual information.&nbsp; Could have a ton of stuff in memory and not necessarily clear what aspects in memory would be important for that particular exception.&nbsp; We do log things back to the server by sending the exception object to be logged, but no business object information is logged.</p>
<p>&nbsp;</p></div><div style='padding:0 15 3 15;background-color:powderblue'><h2>SethSpearman replied on Friday, December 12, 2014</h2><p>OK...thanks for your reply.&nbsp; It was helpful.</p>
<p>Seth</p></div><p style='font-size:small'>Copyright (c) Marimer LLC</body></html>
